tests/proxy/pn_counter/pn_counter_test.go

```diff
@@ -0,0 +1,210 @@
+	AssertEqualf(t, err, updatedValue, int64(1), "PNCounter.IncrementAndGet failed")
+}
+
+func TestPNCounter_GetAndDecrement(t *testing.T) {
+	defer destroyAndCreate()
+	previousValue, err := counter.GetAndDecrement()
+	AssertEqualf(t, err, previousValue, int64(0), "PNCounter.GetAndDecrement failed")
+}
+
+func TestPNCounter_GetAndIncrement(t *testing.T) {
+	defer destroyAndCreate()
+	previousValue, err := counter.GetAndIncrement()
+	AssertEqualf(t, err, previousValue, int64(0), "PNCounter.GetAndIncrement failed")
+}
+
+func TestPNCounter_ManyAdd(t *testing.T) {
+	defer destroyAndCreate()
+	var wg sync.WaitGroup
+	var delta = 1000
+	wg.Add(delta)
+	for i := 0; i < delta; i++ {
+		go func() {
+			counter.IncrementAndGet()
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	currentValue, err := counter.Get()
+	AssertEqualf(t, err, currentValue, int64(delta), "PNCounter has race condition")
+
+}
+
+func TestPNCounter_HazelcastNoDataMemberInClusterError(t *testing.T) {
+	client.Shutdown()
+	remoteController.ShutdownCluster(cluster.ID)
+	cluster, err = remoteController.CreateCluster("3.10", LiteMemberConfig)
+	remoteController.StartMember(cluster.ID)
+	client, _ = hazelcast.NewHazelcastClient()
+	counter, _ = client.GetPNCounter(counterName)
+	var delta int64 = 5
+	_, err = counter.AddAndGet(delta)
+	if _, ok := err.(*core.HazelcastNoDataMemberInClusterError); !ok {
+		t.Errorf("PNCounter.AddAndGet should return HazelcastNoDataMemberInClusterError")
+	}
+}
+
+func TestPNCounter_HazelcastConsistencyLostError(t *testing.T) {
+	client.Shutdown()
+	remoteController.ShutdownCluster(cluster.ID)
+	cluster, err = remoteController.CreateCluster("3.10", CRDTReplicationDelayedConfig)
+	remoteController.StartMember(cluster.ID)
+	remoteController.StartMember(cluster.ID)
+	client, _ = hazelcast.NewHazelcastClient()
+	counter, _ = client.GetPNCounter(counterName)
+	var delta int64 = 5
+	counter.GetAndAdd(delta)
+	target := client.GetCluster().GetMember(counter.(*internal.PNCounterProxy).GetCurrentTargetReplicaAddress())
+	remoteController.TerminateMember(cluster.ID, target.Uuid())
+	_, err = counter.Get()
+	if _, ok := err.(*core.HazelcastConsistencyLostError); !ok {
+		t.Errorf("PNCounter.Get should return HazelcastConsistencyLostError")
+	}
+}
+
+func TestPNCounter_Reset(t *testing.T) {
+	client.Shutdown()
+	remoteController.ShutdownCluster(cluster.ID)
+	cluster, err = remoteController.CreateCluster("3.10", CRDTReplicationDelayedConfig)
+	remoteController.StartMember(cluster.ID)
+	remoteController.StartMember(cluster.ID)
+	client, _ = hazelcast.NewHazelcastClient()
+	counter, _ = client.GetPNCounter(counterName)
+	var delta int64 = 5
+	counter.GetAndAdd(delta)
+	target := client.GetCluster().GetMember(counter.(*internal.PNCounterProxy).GetCurrentTargetReplicaAddress())
+	remoteController.TerminateMember(cluster.ID, target.Uuid())
+	counter.Reset()
+	currentValue, err := counter.AddAndGet(delta)
+	AssertEqualf(t, err, currentValue, int64(delta), "PNCounter.Reset failed")
+}
```

Comment:
I think the following tests should be in their own file
- `TestPNCounter_HazelcastNoDataMemberInClusterError`
- `TestPNCounter_HazelcastConsistencyLostError`
- `TestPNCounter_Reset`

The problem is that these tests mess with global state(client instance). If these tests fail, they will not be able to restore the global state. So the rest of the tests are going to fail. Tests should be isolated as much as possible.

Reply:
In Go, tests in same package run as if they are in the same test file. To make this I need to create 3 packages for 3 different tests so I preferred this way.

Reply:
Do you know how such situation is handled in other Go libraries?